File: src/main/java/com/vk/tlschema/highlighting/TLSchemaAnnotator.java
package com.vk.tlschema.highlighting;

import com.intellij.codeInspection.util.InspectionMessage;
import com.intellij.lang.annotation.AnnotationHolder;
import com.intellij.lang.annotation.Annotator;
import com.intellij.lang.annotation.HighlightSeverity;
import com.intellij.openapi.editor.SyntaxHighlighterColors;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.vk.tlschema.Constants;
import com.vk.tlschema.psi.*;
import com.vk.tlschema.psi.impl.*;
import com.vk.tlschema.search.TLSchemaSearchUtils;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class TLSchemaAnnotator implements Annotator {
    @Override
    public void annotate(@NotNull final PsiElement element, @NotNull AnnotationHolder holder) {
        if (element instanceof TLSchemaResultTypeImpl schemaResultType) {
            ColorType(schemaResultType.getBoxedTypeIdent(), holder, false, true);
            return;
        }

        if (element instanceof TLSchemaFullCombinatorIdImpl fullCombinatorId) {
            TLSchemaLcIdentFull identFull = fullCombinatorId.getLcIdentFull();
            if (identFull != null) {
                TLSchemaLcIdentNs ident = identFull.getLcIdentNs();
                TextRange range = ident.getTextRange();

                setHighlighting(holder, range, TLSchemaSyntaxHighlighter.Constructor);
                if (ident.getTextLength() != identFull.getTextLength()) {
                    TextRange range_num = new TextRange(ident.getTextOffset() + ident.getTextLength(), identFull.getTextOffset() + identFull.getTextLength());
                    setHighlighting(holder, range_num, TLSchemaSyntaxHighlighter.ConstructorHash);
                } else {
                    if (TLSchemaPsiImplUtil.getDeclaration(ident).haveConditionalArgs()) {
                        createWarningAnnotation(holder, range, "Constructor hash should be specified");
                    }
                }
            }
        }

        if (element instanceof TLSchemaTypeTermImpl typeTerm) {
            AnnotateType((TLSchemaTermImpl) typeTerm.getTerm(), holder, false);
        }

        if (element instanceof TLSchemaConditionalDefImpl) {
            TextRange range = element.getTextRange();
            setHighlighting(holder, range, TLSchemaSyntaxHighlighter.FieldsMask);
        }

        if (element instanceof TLSchemaAttributeNodeImpl) {
            PsiElement parent = element.getParent();
            if (parent.getFirstChild() == element) {
                boolean rw_attrs = false;
                for (PsiElement child : parent.getChildren()) {
                    if (child instanceof TLSchemaAttributeNodeImpl) {
                        boolean is_rw = false;
                        for (String name : Constants.rwAnnotations) {
                            if (child.getText().equals(name)) {
                                is_rw = true;
                                break;
                            }
                        }
                        if (is_rw) {
                            if (!rw_attrs) {
                                rw_attrs = true;
                            } else {
                                TextRange range = child.getTextRange();
                                createErrorAnnotation(holder, range, "Only one read/write attribute allowed");
                            }
                        }
                    }
                }
            }
            TextRange range = element.getTextRange();
            boolean known = false;
            for (String name : Constants.validAnnotations) {
                if (element.getText().equals(name)) {
                    known = true;
                    break;
                }
            }

            if (known) {
                setHighlighting(holder, range, TLSchemaSyntaxHighlighter.Attribute);
            } else {
                createErrorAnnotation(holder, range, "Unknown Attribute");
            }
        }

    }

    private void AnnotateType(TLSchemaTermImpl element, @NotNull AnnotationHolder holder, boolean percent) {

        if (element.getExpr() != null) {
            TLSchemaExprImpl expr = (TLSchemaExprImpl) element.getExpr();
            for (TLSchemaTerm t : expr.getTermList()) {
                AnnotateType((TLSchemaTermImpl) t, holder, percent);
                percent = false;
            }
            return;
        }

        if (element.getNatExpr() != null) {
            TLSchemaNatExpr expr = element.getNatExpr();
            TextRange range = expr.getTextRange();
            setHighlighting(holder, range, SyntaxHighlighterColors.NUMBER);
            if (percent) {
                createErrorAnnotation(holder, range, "Percent can't be applied to number");
            }

            return;
        }

        if (element.getPercentTerm() != null) {
            if (percent) {
                TextRange range = element.getTextRange();
                createErrorAnnotation(holder, range, "Double percent");
                return;
            }
            AnnotateType((TLSchemaTermImpl) element.getPercentTerm().getTerm(), holder, true);
            return;
        }

        if (element.getTypeIdent() != null) {
            ColorType(element.getTypeIdent(), holder, percent, false);
            return;
        }

        if (element.getVarIdent() != null) {
            throw new AssertionError();
        }

        if (element.getTypeWithTriangleBraces() != null) {
            TLSchemaTypeWithTriangleBraces type = element.getTypeWithTriangleBraces();
            TLSchemaTypeIdent ident = type.getTypeIdent();
            ColorType(ident, holder, percent, false);
            for (TLSchemaTypeExpr e : type.getTypeExprList()) {
                for (TLSchemaTypeTerm t : e.getTypeTermList()) {
                    AnnotateType((TLSchemaTermImpl) t.getTerm(), holder, false);
                }
            }
            return;
        }

        throw new AssertionError();
    }

    private void ColorType(@NotNull PsiElement ident, @NotNull AnnotationHolder holder, boolean percent, boolean no_bare) {
        String type = ident.getText();
        TextRange range = ident.getTextRange();
        boolean bare = false;
        TLSchemaLcIdentNs combinator = TLSchemaSearchUtils.findCombinator(
                ident.getProject(),
                ident.getContainingFile(),
                type
        );
        if (combinator != null) {
            bare = true;
            type = combinator.getDeclaration().getResultType().getBoxedTypeIdent().getText();
        }
        List<TLSchemaResultType> cons = TLSchemaSearchUtils.findType(ident.getProject(), ident.getContainingFile(), type);
        boolean is_simple_type = cons.size() == 1;
        if (is_simple_type) {
            TLSchemaResultType resultType = cons.get(0);
            PsiElement parent = resultType.getParent();
            if (parent instanceof TLSchemaCombinatorDecl combinatorDecl) {
                if (combinatorDecl.getFullCombinatorId().getLcIdentFull() != null) {
                    String name = combinatorDecl.getFullCombinatorId().getLcIdentFull().getLcIdentNs().getName();
                    String type_name = resultType.getBoxedTypeIdent().getUcIdentNs().getName();
                    if (name != null && type_name != null && !name.equalsIgnoreCase(type_name)) {
                        is_simple_type = false;
                    }
                }
            }
        }
        TLSchemaDeclaration decl = TLSchemaPsiImplUtil.getDeclaration(ident);
        if (decl == null) throw new AssertionError();
        List<TLSchemaVarIdent> numVars = decl.getNumVars();
        if (numVars != null) {
            for (TLSchemaVarIdent id : numVars) {
                if (id.getText().equals(type)) {
                    setHighlighting(holder, range, TLSchemaSyntaxHighlighter.NumericVar);
                    return;
                }
            }
        }
        boolean type_arg = false;
        List<TLSchemaVarIdent> typeVars = decl.getTypeVars();
        if (typeVars != null) {
            for (TLSchemaVarIdent id : typeVars) {
                if (id.getText().equals(type)) {
                    type_arg = true;
                    break;
                }
            }
        }
        if (cons.isEmpty() && !type.equals("#") && !type.equals("Type") && !type_arg) {
            createErrorAnnotation(holder, range, "Unknown type");
        } else if (is_simple_type && !percent && !bare && !no_bare) {
            createWarningAnnotation(holder, range, "This type can be made bare");
        } else if ((!is_simple_type || no_bare || type_arg) && (percent || bare)) {
            createErrorAnnotation(holder, range, "This type can't be made bare");
        } else {
            if (percent || bare) {
                setHighlighting(holder, range, TLSchemaSyntaxHighlighter.BareType);
            } else {
                setHighlighting(holder, range, TLSchemaSyntaxHighlighter.BoxedType);
            }
        }
    }

    private static void setHighlighting(
            @NotNull AnnotationHolder holder,
            @NotNull TextRange range,
            @NotNull TextAttributesKey key) {
        holder.newSilentAnnotation(HighlightSeverity.INFORMATION).range(range).textAttributes(key).create();
    }

    private static void createWarningAnnotation(
            @NotNull AnnotationHolder holder,
            @NotNull TextRange range,
            @NotNull @InspectionMessage String message) {
        holder.newAnnotation(HighlightSeverity.WARNING, message).range(range).create();
    }

    private static void createErrorAnnotation(
            @NotNull AnnotationHolder holder,
            @NotNull TextRange range,
            @NotNull @InspectionMessage String message) {
        holder.newAnnotation(HighlightSeverity.ERROR, message).range(range).create();
    }
}

File: src/main/kotlin/com/vk/tlschema/highlighting/TLSchemaAnnotator.kt
package com.vk.tlschema.highlighting

import com.intellij.lang.annotation.Annotation
import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.openapi.editor.SyntaxHighlighterColors
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.vk.tlschema.Constants
import com.vk.tlschema.psi.TLSchemaCombinatorDecl
import com.vk.tlschema.psi.impl.*
import com.vk.tlschema.search.TLSchemaSearchUtils
import java.util.*

class TLSchemaAnnotator : Annotator {
    override fun annotate(element: PsiElement, holder: AnnotationHolder) {
        if (element is TLSchemaResultTypeImpl) {
            ColorType(element.boxedTypeIdent, holder, false, true)
            return
        }
        if (element is TLSchemaFullCombinatorIdImpl) {
            val identFull = element.lcIdentFull
            if (identFull != null) {
                val ident = identFull.lcIdentNs
                val range = ident.textRange
                var annotation = holder.createInfoAnnotation(range, null)
                annotation.textAttributes = TLSchemaSyntaxHighlighter.Constructor
                if (ident.textLength != identFull.textLength) {
                    val range_num =
                        TextRange(ident.textOffset + ident.textLength, identFull.textOffset + identFull.textLength)
                    val annotation_num = holder.createInfoAnnotation(range_num, null)
                    annotation_num.textAttributes = TLSchemaSyntaxHighlighter.ConstructorHash
                } else {
                    if (TLSchemaPsiImplUtil.getDeclaration(ident).haveConditionalArgs()) {
                        annotation = holder.createWarningAnnotation(range, "Constructor hash should be specified")
                    }
                }
            }
        }
        if (element is TLSchemaTypeTermImpl) {
            AnnotateType(element.term as TLSchemaTermImpl, holder, false)
        }
        if (element is TLSchemaConditionalDefImpl) {
            val range = element.getTextRange()
            val annotation = holder.createInfoAnnotation(range, null)
            annotation.textAttributes = TLSchemaSyntaxHighlighter.FieldsMask
        }
        if (element is TLSchemaAttributeNodeImpl) {
            val parent = element.getParent()
            if (parent.firstChild === element) {
                var rw_attrs = false
                for (child in parent.children) {
                    if (child is TLSchemaAttributeNodeImpl) {
                        var is_rw = false
                        for (name in Constants.rwAnnotations) {
                            if (child.getText() == name) {
                                is_rw = true
                                break
                            }
                        }
                        if (is_rw) {
                            if (!rw_attrs) {
                                rw_attrs = true
                            } else {
                                holder.createErrorAnnotation(
                                    child.getTextRange(),
                                    "Only one read/write attribute allowed"
                                )
                            }
                        }
                    }
                }
            }
            val range = element.getTextRange()
            val annotation: Annotation
            var known = false
            for (name in Constants.validAnnotations) {
                if (element.getText() == name) {
                    known = true
                    break
                }
            }
            annotation = if (known) holder.createInfoAnnotation(range, null) else holder.createErrorAnnotation(
                range,
                "Unknown Attribute"
            )
            if (known) {
                annotation.textAttributes = TLSchemaSyntaxHighlighter.Attribute
            }
        }
    }

    private fun AnnotateType(element: TLSchemaTermImpl, holder: AnnotationHolder, percent: Boolean) {
        var percent = percent
        if (element.expr != null) {
            val expr = element.expr as TLSchemaExprImpl?
            for (t in expr!!.termList) {
                AnnotateType(t as TLSchemaTermImpl, holder, percent)
                percent = false
            }
            return
        }

        if (element.natExpr != null) {
            val expr = element.natExpr
            val annotation: Annotation
            val range = expr!!.textRange
            annotation = if (percent) {
                holder.createErrorAnnotation(range, "Percent can't be applied to number")
            } else {
                holder.createInfoAnnotation(range, null)
            }
            annotation.textAttributes = SyntaxHighlighterColors.NUMBER
            return
        }

        if (element.percentTerm != null) {
            if (percent) {
                val range = element.textRange
                holder.createErrorAnnotation(range, "Double percent")
                return
            }
            AnnotateType(element.percentTerm!!.term as TLSchemaTermImpl, holder, true)
            return
        }

        if (element.typeIdent != null) {
            ColorType(element.typeIdent!!, holder, percent, false)
            return
        }

        if (element.varIdent != null) {
            throw AssertionError()
        }

        if (element.typeWithTriangleBraces != null) {
            val type = element.typeWithTriangleBraces
            val ident = type!!.typeIdent
            ColorType(ident, holder, percent, false)
            for (e in type.typeExprList) {
                for (t in e.typeTermList) {
                    AnnotateType(t.term as TLSchemaTermImpl, holder, false)
                }
            }
            return
        }

        throw AssertionError()
    }

    private fun ColorType(ident: PsiElement, holder: AnnotationHolder, percent: Boolean, no_bare: Boolean) {
        var type = ident.text
        val range = ident.textRange
        var bare = false
        val combinator = TLSchemaSearchUtils.findCombinator(ident.project, type)
        if (combinator != null) {
            bare = true
            type = combinator.declaration.resultType.boxedTypeIdent.text
        }
        val cons = TLSchemaSearchUtils.findType(ident.project, type)
        var is_simple_type = cons.size == 1
        if (is_simple_type) {
            val resultType = cons[0]
            val parent = resultType.parent
            if (parent is TLSchemaCombinatorDecl) {
                val decl = parent
                if (decl.fullCombinatorId.lcIdentFull != null) {
                    val name = decl.fullCombinatorId.lcIdentFull!!.lcIdentNs.name
                    val type_name = resultType.boxedTypeIdent.ucIdentNs.name
                    if (name != null && type_name != null && name.lowercase(Locale.getDefault()) != type_name.lowercase(
                            Locale.getDefault()
                        )
                    ) {
                        is_simple_type = false
                    }
                }
            }
        }
        val decl = TLSchemaPsiImplUtil.getDeclaration(ident) ?: throw AssertionError()
        val numVars = decl.numVars
        if (numVars != null) {
            for (id in numVars) {
                if (id.text == type) {
                    val annotation = holder.createInfoAnnotation(range, null)
                    annotation.textAttributes = TLSchemaSyntaxHighlighter.NumericVar
                    return
                }
            }
        }
        var type_arg = false
        val typeVars = decl.typeVars
        if (typeVars != null) {
            for (id in typeVars) {
                if (id.text == type) {
                    type_arg = true
                }
            }
        }
        if (cons.size == 0 && type != "#" && type != "Type" && !type_arg) {
            holder.createErrorAnnotation(range, "Unknown type")
        } else if (is_simple_type && !percent && !bare && !no_bare) {
            holder.createWarningAnnotation(range, "This type can be made bare")
        } else if ((!is_simple_type || no_bare || type_arg) && (percent || bare)) {
            holder.createErrorAnnotation(range, "This type can't be made bare")
        } else {
            val annotation = holder.createInfoAnnotation(range, null)
            if (percent || bare) {
                annotation.textAttributes = TLSchemaSyntaxHighlighter.BareType
            } else {
                annotation.textAttributes = TLSchemaSyntaxHighlighter.BoxedType
            }
        }
    }
}
